edb/server/main.py

```diff
@@ -609,19 +609,31 @@ def server_main(**kwargs):
             asyncio.run(run_server(server_args))
 
 
-@click.command(
+@click.group(
     'EdgeDB Server',
+    invoke_without_command=True,
     context_settings=dict(help_option_names=['-h', '--help']))
 @srvargs.server_options
-def main(version=False, **kwargs):
+@click.pass_context
+def main(ctx, version=False, **kwargs):
     if kwargs.get('testmode') and 'EDGEDB_TEST_CATALOG_VERSION' in os.environ:
         buildmeta.EDGEDB_CATALOG_VERSION = int(
             os.environ['EDGEDB_TEST_CATALOG_VERSION']
         )
     if version:
         print(f"edgedb-server, version {buildmeta.get_version()}")
         sys.exit(0)
-    server_main(**kwargs)
+    logsetup.setup_logging(kwargs['log_level'], kwargs['log_to'])
+    if ctx.invoked_subcommand is None:
+        server_main(**kwargs)
+
+
+@main.command()
+@srvargs.compiler_options
+def compiler(**kwargs):
+    from edb.server.compiler_pool import server as compiler_server
+
+    asyncio.run(compiler_server.server_main(**kwargs))
```

Comment:
Does this have to be defined in this file? Maybe `compiler_main.py`?

Reply:
I usually use a `entry_point` in `setup.py` to inject the dependencies of the CLI modules (example [here](https://python-gino.org/docs/en/1.0/tutorials/fastapi.html#create-models-and-api) using `importlib` and Poetry plugins), in order to avoid a late import (which is necessary to avoid a circular import) like this:

`main.py`:

```python
@click.group(...)
def main(...):
    ...

from . import compiler_main  # noqa: late import to avoid circular import
```

`compiler_main.py`:

```python
from . import main

@main.command()
def compiler(...):
    from edb.server.compiler_pool import server
    ...server.server_main()...
```

For the EdgeDB server, I don't think we need a dependency injection because we're not a extensible library. So I was leaning towards just putting the `compiler` subcommand in `edb.server.main`, because a late import cannot hide too much implementation details (like 1 line vs 5 lines, and we still see the "compiler" keyword in the main server).

Even with that said, I'm still fine with adding the `from . import compiler_main` here, just reply 🚀 ;)

Reply:
I was asking why not add another script "edgedb_compiler_pool" to https://github.com/edgedb/edgedb/blob/f4c1ced364a94e8206297fb970ef494c9b1a2de3/setup.py#L820-L824 and keep everything related in a separate file from main.py?

Reply:
Ohhhh I see what you meant. I think @elprans had some quick thoughts to use a subcommand under `edgedb-server` (to minimize the number of executables?).

Reply:
Yeah. Exposing entrypoints in prod builds is a pain, so it would be best take the "single entrypoint" approach to operational modes.